src/transformStateWithClones.js

```diff
@@ -7,7 +7,43 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const copyOfState = { ...state };
+  const stateCopies = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(copyOfState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeObjectProperties(action.keysToRemove, copyOfState);
+        break;
+
+      case 'clear':
+        clearObject(copyOfState);
+        break;
+
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    stateCopies.push({ ...copyOfState });
+  }
+
+  return stateCopies;
+}
+
+function clearObject(object) {
+  for (const key in object) {
+    delete object[key];
+  }
+}
+
+function removeObjectProperties(propertiesObject, objectToModify) {
+  for (const key of propertiesObject) {
+    delete objectToModify[key];
+  }
 }
```

Comment:
usually it is considered to be a bad practice to use data types in your variables/function names.
As you have it quite abstract here (it looks like you can pass here any object and it will reset it) - it is ok in this context. But consider it in your next Pull Requests